### PR TITLE
Specified generic version for Octokit defaults

### LINKED ISSUE #11: Make the `version` parameter work with the static `.withDefaults()` method
In a nutshell I want `octokitGhes31_v1` and `octokitGhes31_v2` to have both the internal version type set to `"ghes-3.1."`

```ts
const OctokitWithDefaultVersion = Octokit.withDefaults({ version: "ghes-3.1" });

const octokitGhes31_v1 = new OctokitWithDefaultVersion();
const octokitGhes31_v2 = new Octokit({ version: "ghes-3.1" });
```

Here is a TypeScript test file that should compile, but currently has several errors

```ts
import { Octokit } from "@octokit-next/core";

import "@octokit-next/types-rest-api-ghes-3.0";

export async function run() {
  const octokit = new Octokit({
    version: "ghes-3.1",
  });

  const response = await octokit.request("GET /ghes-only");

  expectType<boolean>(response.data.ok);

  const OctokitGHES31 = Octokit.withDefaults({
    version: "ghes-3.1",
  });
  const octokitGhes31 = new OctokitGHES31();
  expectType<never>(await octokitGhes31.request("GET /dotcom-only"));

  const octokitGhes30ViaConstructorOptions = new OctokitGHES31({
    version: "ghes-3.0",
  });
  expectType<never>(
    await octokitGhes30ViaConstructorOptions.request("GET /new-ghes-only")
  );

  const OctokitGHES30 = OctokitGHES31.withDefaults({
    version: "ghes-3.0",
  });
  const octokitGhes30ViaChainedDefaults = new OctokitGHES30();
  expectType<never>(
    await octokitGhes30ViaChainedDefaults.request("GET /new-ghes-only")
  );
}

function expectType<T>(value: T): void {}
```

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Fixes #11.

The key thing was that in order to know `version` to be a specific (generic) one of the potential API versions. That has to be specified per `Options<TVersion>` ... which then spreads to wherever a new `Options` is set up, including constructors that take it in.

### CODE REVIEW

File: packages/types/versions.test-d.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    const octokit = new Octokit({
+        version: "ghes-3.1",
+    });
+
+    const response = await octokit.request("GET /ghes-only");
+
+    expectType<boolean>(response.data.ok);
+
+    const OctokitGHES31 = Octokit.withDefaults({
+        version: "ghes-3.1",
```
Comment (JoshuaKGoldberg) [inline packages/types/versions.test-d.ts:16]: A fun little tip for debugging: when mouse-hovering over the `version` in each of these calls, if it comes up as `string` then TypeScript isn't being told it needs to specify the generic for the function call. If it comes up as the string literal such as `"ghes-3.1"` here then TypeScript is storing it as the specific literal.

**Comment by JoshuaKGoldberg** [comment]
Aha, there is a test failure after all:

```
  tests/js/request/index.test-d.ts:8:33
  ✖  8:33  No overload matches this call.
```

Working on it...

**Comment by gr2m** [comment]
The tsd tests in `packages/*/` are currently not yet executed, only the ones in `tests/js/*`. I'll fix that probably today but for now I'll move all tests to `tests/js/*` to make sure that they are all run with `npm test`

**Comment by gr2m** [comment]
Hmm when I run `npx tsd tests/js/plugins-and-defaults` shouldn't that evaluate `tests/js/plugins-and-defaults/versions.test-d.ts`? It doesn't seem like it does that right now, even when I put dummy `version.js` and `version.d.ts` files into the folder. I added `expectType<number>("string");` to `versions.test-d.ts` and running `npx tsd tests/js/plugins-and-defaults` does not complain

**Comment by gr2m** [comment]
hm nevermind my last push that mixes up the types coming from plugins which requires a `required` option

**Comment by gr2m** [comment]
I thin the current error 

> ```
> tests/js/with-defaults-version/index.test-d.ts:20:2
> ✖  20:2  Parameter type Response<any, any> is not identical to argument type Response<{ "+1": string; "-1": string; "dotcom-only": string; }, DotcomResponseHeaders> & Response<{ "+1": string; "-1": string; "ghes-only": string; "dotcom-only": never; }, GHES31ResponseHeaders>.
> ```

Is caused by how I create the types for GHES 3.1 in 

https://github.com/octokit/octokit-next.js/blob/18c1abcd291c202aa6dd1f349bb1297b7fc4cbaf/packages/types-rest-api-ghes-3.1/index.d.ts#L58-L59

I think the correct way to do it is to only use the Endpoints types from `Octokit.ApiVersions["github.com"]["Endpoints"]` that do not exist in `GHES30EndpointsDiff`, instead of naively creating an intersection?

**Comment by gr2m** [comment]
I'm not sure if it's `tsd` or the local configuration, but it looks like we cannot use `exepecType<any>(...)`, it fails with anything but `expectType<any>({} as any);`

**Review by gr2m — APPROVED** [review]
This is great, thank you so much, I learned a lot!

I was able to simplify the tests for `octokit.options` via https://github.com/octokit/octokit-next.js/pull/24/commits/e6292a701fa6ccb83e6ddc49ace62b0aa0b983d0#diff-144904d408ec513806681518c894e534492562fdd352bba77ab8e7643851d978R255, and I added a few more tests.

Thanks so much for making this work!

**Comment by JoshuaKGoldberg** [comment]
Great! I'm just now catching up on the thread, glad it's all working!

**Pull request merged**